include/dali/core/tensor_layout.h

```diff
@@ -0,0 +1,475 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_TENSOR_LAYOUT_H_
+#define DALI_CORE_TENSOR_LAYOUT_H_
+
+#include <cassert>
+#include <cstring>
+#include <array>
+#include <string>
+#include <stdexcept>
+#include "dali/core/error_handling.h"
+#include "dali/core/host_dev.h"
+
+namespace dali {
+
+/** @brief Provides a domain-agnostic, flexible description of a tensor layout
+ *
+ * The object is essentially a string with storage optimized for short sequences.
+ */
+class TensorLayout {
+ public:
+  DALI_HOST_DEV
+  constexpr TensorLayout() { set_size(0); }
+
+  /** @brief Constructs a TensorLayout from a C-style string */
+  DALI_HOST_DEV
+  constexpr TensorLayout(const char *str) {  // NOLINT
+    int i = 0;
+    for (; str[i] && i < max_ndim; i++) {
+      data_[i] = str[i];
+    }
+    int n = i;
+    for (; i < max_ndim; i++) {
+      data_[i] = 0;
+    }
+    set_size(n);
+  }
+
+  /** @brief Constructs a TensorLayout from a C-style string of known length */
+  DALI_HOST_DEV
+  constexpr TensorLayout(const char *str, size_t n) {  // NOLINT
+    assert(n < sizeof(data_));
+    if (n >= sizeof(data_))
+      n = sizeof(data_) - 1;
+    for (size_t i = 0; i < n; i++)
+      data_[i] = str[i];
+    data_[n] = 0;
+    set_size(n);
+  }
+
+  /** @brief Constructs a TensorLayout from a char array of known length, e.g. a string literal */
+  template <size_t N>
+  DALI_HOST_DEV
+  constexpr TensorLayout(const char (&s)[N])  // NOLINT
+  : TensorLayout(s, N && s[N-1] == '\0' ? N - 1 : N) {
```

Comment:
Before you were using `0`. Maybe update the previous occurrences to use `\0` as well